SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialEventTracer.cpp

```diff
@@ -74,89 +74,184 @@ SpatialEventTracer::~SpatialEventTracer()
 	}
-	return HexStr;
-}
 
-FUserSpanId SpatialEventTracer::SpanIdToUserSpanId(const Trace_SpanId& SpanId)
-{
 	FUserSpanId UserSpanId;
-	UserSpanId.Data.SetNum(TraceSpanIdLength);
-	const int32 Size = TraceSpanIdLength * sizeof(uint8);
-	FMemory::Memcpy(UserSpanId.Data.GetData(), SpanId.data, Size);
+	UserSpanId.Data.SetNum(TRACE_SPAN_ID_SIZE_BYTES);
+	const int32 Size = TRACE_SPAN_ID_SIZE_BYTES * sizeof(uint8);
+	FMemory::Memcpy(UserSpanId.Data.GetData(), SpanId.GetConstData(), Size);
 	return UserSpanId;
 }
 
-TOptional<Trace_SpanId> SpatialEventTracer::UserSpanIdToSpanId(const FUserSpanId& UserSpanId)
+FSpatialGDKSpanId SpatialEventTracer::UserSpanIdToGDKSpanId(const FUserSpanId& UserSpanId)
 {
 	if (!UserSpanId.IsValid())
 	{
-		return {};
+		return FSpatialGDKSpanId(false);
 	}
 
-	Trace_SpanId SpanId;
-	const int32 Size = TraceSpanIdLength * sizeof(uint8);
-	FMemory::Memcpy(SpanId.data, UserSpanId.Data.GetData(), Size);
-	return SpanId;
+	FSpatialGDKSpanId TraceSpanId(true);
+	const int32 Size = TRACE_SPAN_ID_SIZE_BYTES * sizeof(uint8);
+	FMemory::Memcpy(TraceSpanId.GetData(), UserSpanId.Data.GetData(), Size);
+	return TraceSpanId;
 }
 
-TOptional<Trace_SpanId> SpatialEventTracer::CreateSpan() const
+FSpatialGDKSpanId SpatialEventTracer::CreateSpan(const Trace_SpanIdType* Causes /* = nullptr*/, int32 NumCauses /* = 0*/) const
 {
 	if (!IsEnabled())
 	{
-		return {};
+		return FSpatialGDKSpanId(false);
+	}
+
+	if (Causes == nullptr && NumCauses > 0)
+	{
+		return FSpatialGDKSpanId(false);
+	}
+
+	FSpatialGDKSpanId TraceSpanId(true);
+
+	Trace_SamplingResult SamplingResult = Trace_EventTracer_ShouldSampleSpan(EventTracer, Causes, NumCauses, nullptr);
+	if (SamplingResult.decision == Trace_SamplingDecision::TRACE_SHOULD_NOT_SAMPLE)
+	{
+		return FSpatialGDKSpanId(false);
+	}
+
+	if (Causes != nullptr && NumCauses > 0)
+	{
+		Trace_EventTracer_AddSpan(EventTracer, Causes, NumCauses, nullptr, TraceSpanId.GetData());
+	}
+	else
+	{
+		Trace_EventTracer_AddSpan(EventTracer, Trace_SpanId_Null(), 1, nullptr, TraceSpanId.GetData());
 	}
 
-	return Trace_EventTracer_AddSpan(EventTracer, nullptr, 0);
+	return TraceSpanId;
 }
 
-TOptional<Trace_SpanId> SpatialEventTracer::CreateSpan(const Trace_SpanId* Causes, int32 NumCauses) const
+void SpatialEventTracer::TraceEvent(const FSpatialTraceEvent& SpatialTraceEvent, const FSpatialGDKSpanId& SpanId)
 {
 	if (!IsEnabled())
 	{
-		return {};
+		return;
 	}
 
-	if (Causes != nullptr && NumCauses > 0)
+	if (!SpanId.IsValid())
 	{
-		return Trace_EventTracer_AddSpan(EventTracer, Causes, NumCauses);
+		return;
 	}
 
-	return Trace_EventTracer_AddSpan(EventTracer, nullptr, 0);
+	auto MessageSrc = StringCast<ANSICHAR>(*SpatialTraceEvent.Message);
+	auto TypeSrc = StringCast<ANSICHAR>(*SpatialTraceEvent.Type.ToString());
+
+	Trace_Event Event;
+	Event.span_id = SpanId.GetConstData();
+	Event.message = MessageSrc.Get();
+	Event.type = TypeSrc.Get();
+	Event.unix_timestamp_millis = 0;
+	Event.data = nullptr;
```

Comment:
Same here (as below)
```suggestion
	Trace_Event Event{};
	Event.span_id = SpanId.GetConstData();
	Event.message = MessageSrc.Get();
	Event.type = TypeSrc.Get();
```